container/src/main/java/com/flipkart/poseidon/core/PoseidonServlet.java

```diff
@@ -245,24 +249,25 @@ private void internalError(HttpServletResponse httpResponse, Throwable throwable
     }
 
     private void processErrorResponse(int statusCode, HttpServletResponse httpResponse, Throwable throwable) throws IOException {
-        MediaType contentType = application.getDefaultMediaType();
-        String errorMsg = "";
-        Throwable generatedException = ExceptionUtils.getRootCause(throwable);
-
-        if (generatedException != null && generatedException instanceof DataSourceException) {
-            DataSourceException dsException = (DataSourceException) generatedException;
-            if (dsException.getResponse() != null) {
-                errorMsg = ObjectMapperHelper.getMapper().writeValueAsString(dsException.getResponse());
-            }
-            if (dsException.getStatusCode() > 0) {
-                statusCode = dsException.getStatusCode();
+        if (configuration.getExceptionMapper() == null || !configuration.getExceptionMapper().map(Optional.ofNullable(ExceptionUtils.getRootCause(throwable)).orElse(throwable), httpResponse)) {
```

Comment:
Too much of nested calls. Please see if this can be broken. Also note we would be calling ExceptionUtils.getRootCause() twice when the mapper returns false (it couldn't map exception to a http response). This should and could be avoided.